server/server.go

```diff
@@ -6,50 +6,185 @@ package server
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:      id.String(),
+		logger:    logp.NewLogger("shipper-server"),
+		publisher: publisher,
+		close:     &sync.Once{},
+	}
+
+	s.ctx, s.stop = context.WithCancel(context.Background())
+
+	return &s, nil
+}
+
+// GetAcceptedIndex returns the accepted index
+func (serv *shipperServer) GetAcceptedIndex() uint64 {
+	return uint64(serv.publisher.AcceptedIndex())
+}
+
+// GetPersistedIndex returns the persisted index
+func (serv *shipperServer) GetPersistedIndex() uint64 {
+	return uint64(serv.publisher.PersistedIndex())
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
+	resp := &messages.PublishReply{
+		Uuid: serv.uuid,
+	}
+
+	// the value in the request is optional
+	if req.Uuid != "" && req.Uuid != serv.uuid {
+		resp.AcceptedIndex = serv.GetAcceptedIndex()
+		resp.PersistedIndex = serv.GetPersistedIndex()
+		serv.logger.Debugf("shipper UUID does not match, all events rejected. Expected = %s, actual = %s", serv.uuid, req.Uuid)
+
+		return resp, nil
+	}
+
+	for _, e := range req.Events {
+		_, err := serv.publisher.Publish(e)
+		if err == nil {
+			resp.AcceptedCount++
+			continue
+		}
+
+		if errors.Is(err, queue.ErrQueueIsFull) {
+			serv.logger.Debugf("queue is full, not all events accepted. Events = %d, accepted = %d", len(req.Events), resp.AcceptedCount)
+		} else {
+			err = fmt.Errorf("failed to enqueue an event. Events = %d, accepted = %d: %w", len(req.Events), resp.AcceptedCount, err)
+			serv.logger.Error(err)
+		}
+
+		break
+	}
+
+	resp.AcceptedIndex = serv.GetAcceptedIndex()
+	resp.PersistedIndex = serv.GetPersistedIndex()
+
+	serv.logger.
+		Debugf("finished publishing a batch. Events = %d, accepted = %d, accepted index = %d, persisted index = %d",
+			len(req.Events),
+			resp.AcceptedCount,
+			resp.AcceptedIndex,
+			resp.PersistedIndex,
+		)
+
+	return resp, nil
+}
+
+// PublishEvents is the server implementation of the gRPC PersistedIndex call.
+func (serv *shipperServer) PersistedIndex(req *messages.PersistedIndexRequest, producer pb.Producer_PersistedIndexServer) error {
+	serv.logger.Debug("new subscriber for persisted index change")
+	defer serv.logger.Debug("unsubscribed from persisted index change")
+
+	persistedIndex := serv.GetPersistedIndex()
+	if persistedIndex != 0 {
+		err := producer.Send(&messages.PersistedIndexReply{
+			Uuid:           serv.uuid,
+			PersistedIndex: persistedIndex,
+		})
 		if err != nil {
-			// If we couldn't accept any events, return the error directly. Otherwise,
-			// just return success on however many events we were able to handle.
-			if reply.AcceptedCount == 0 {
-				return nil, err
+			return err
+		}
+	}
+
+	pollingIntervalDur := req.PollingInterval.AsDuration()
+
+	if pollingIntervalDur == 0 {
+		return nil
```

Comment:
It feels a bit weird not to send a response in this case but it does make sense for a stream. I guess if this use case is common we can just add a unary RPC for convenience.

Reply:
We send a response, please see above. As I stated in the documentation if the interval is zero we don't subscribe the client but rather send a single response and close the stream https://github.com/elastic/elastic-agent-shipper-client/blob/ce5fac5d4177ece459766ff56d39028ed28da937/api/messages/persisted_index.proto#L16-L20